server/notebooks/api_views.py

```diff
@@ -80,9 +81,29 @@ def perform_destroy(self, instance):
 
     def perform_create(self, serializer):
         ctx = self.get_serializer_context()
-        notebook_owner_id = Notebook.objects.values_list("owner", flat=True).get(
-            id=ctx["notebook_id"]
-        )
-        if self.request.user.id != notebook_owner_id:
+        notebook = Notebook.objects.get(id=ctx["notebook_id"])
+
+        # check if user has permission to create a revision
+        if self.request.user.id != notebook.owner_id:
             raise PermissionDenied
+
+        # check that this revision has genuinely new content
+        last_revision = NotebookRevision.objects.filter(notebook_id=notebook.id).first()
+        if (
+            serializer.validated_data["title"] == last_revision.title
+            and serializer.validated_data["content"] == last_revision.content
+        ):
+            return (False, NotebookRevisionDetailSerializer(last_revision).data)
```

Comment:
This smells like something that could live in the validation step of the serializer to me? I'm imagining there's no case where we'd want to save a revision that has no change. Also as a validation step the API can return a non-200 response (a 4xx) that can then be used to display something in the UI. In other words, 200 is a successful save but a 400 is not.

Reply:
Hmm, you might be right about this now that I think of it. When I wrote this I figured we would simply "accept" revisions that were unchanged and do nothing with them, but perhaps a 400 code would be better. Let me investigate that approach, we may need to make some minor modifications on the client for this...

Reply:
Your proposed approach has the additional advantage of not requiring so much overloading of the view.

Reply:
Yes, this is better! Pushed a new revision with the suggested change.